pkg/controller/machine/drain_controller.go

```diff
@@ -182,3 +191,41 @@ func (d *machineDrainController) drainNode(ctx context.Context, machine *machine
 
 	return nil
 }
+
+// isDrainAllowed checks whether the drain is permitted at this time.
+// It checks the following:
+// - Is the node cordoned, if so allow draining to complete any previous attempt to drain.
+// - Is the node a control plane node, if so, only allow draining if no other control plane node is already being drained.
+func (d *machineDrainController) isDrainAllowed(ctx context.Context, node *corev1.Node) error {
+	if node.Spec.Unschedulable {
+		// If the node has already been cordoned, continue to drain.
+		return nil
+	}
+
+	if !isControlPlaneNode(*node) {
+		// We always allow draining of worker nodes.
+		return nil
+	}
+
+	nodes := &corev1.NodeList{}
+	if err := d.Client.List(ctx, nodes); err != nil {
+		return fmt.Errorf("could not list control plane nodes: %v", err)
+	}
+
+	for _, otherNode := range nodes.Items {
+		if isControlPlaneNode(otherNode) && otherNode.Spec.Unschedulable {
```

Comment:
What would happen in a situation when drain on a master node reached timeout a was requeued? Looking at the code, i have feeling that it would prohibit drain because this node was cordoned on previous iteration.

Reply:
It prohibits us from draining any other node, but when that one comes back up, we will allow it to schedule because of the `node.Spec.Unschedulable` shortcut at the top of this function

Reply:
Gotcha, missed that, thx